### PR TITLE
:bug: Improve MHC validation for topology-managed MHC

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Add validation for nodeStartupTimeout on cluster creation. Merge MHC validation into a single function in the MHC api definition which can then be used for common validation no matter if the MHC is defined in the Cluster topology, the ClusterClass or an individuation MHC object.

### CODE REVIEW

File: api/v1beta1/machinehealthcheck_webhook.go [hunk 1]
```diff
@@ -165,8 +175,12 @@ func (m *MachineHealthCheck) validate(old *MachineHealthCheck) error {
 		)
 	}
 
-	if len(allErrs) == 0 {
-		return nil
+	if len(m.Spec.UnhealthyConditions) == 0 {
```
Comment (killianmuldoon) [inline api/v1beta1/machinehealthcheck_webhook.go:178]: This validation could be left up to kubebuilder, which is how we're doing it for MHCSpec today.

File: internal/webhooks/cluster.go [hunk 2]
```diff
@@ -366,65 +367,56 @@ func (webhook *Cluster) getClusterClassForCluster(ctx context.Context, cluster *
...
-	if cluster.Spec.Topology.ControlPlane.MachineHealthCheck != nil &&
-		cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable != nil &&
-		*cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable &&
-		cluster.Spec.Topology.ControlPlane.MachineHealthCheck.MachineHealthCheckClass.IsZero() &&
-		clusterClass.Spec.ControlPlane.MachineHealthCheck == nil {
-		allErrs = append(allErrs, field.Forbidden(
-			field.NewPath("spec", "topology", "controlPlane", "machineHealthCheck", "enable"),
-			fmt.Sprintf("cannot be set to %t as MachineHealthCheck definition is not available in the Cluster topology or the ClusterClass", *cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable),
-		))
+		allErrs = append(allErrs, validateMachineHealthCheckTopology(path, cluster.Namespace, cluster.Spec.Topology.ControlPlane.MachineHealthCheck)...)
```
Comment (killianmuldoon) [inline internal/webhooks/cluster.go:396]: Trickiest section here is trying to correctly merge the MHC Toplogy with the underlying MHC Class to get a validatable result here.

File: api/v1beta1/machinehealthcheck_webhook.go [hunk 3]
```diff
@@ -136,15 +136,27 @@ func (m *MachineHealthCheck) validate(old *MachineHealthCheck) error {
...
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("MachineHealthCheck").GroupKind(), m.Name, allErrs)
+}
+
+// ValidateCommonFields validates UnhealthyConditions NodeStartupTimeout, MaxUnhealthy, and RemediationTemplate of the MHC.
+// It is used to help validate other types which define MachineHealthChecks such as MachineHealthCheckClass and MachineHealthCheckTopology.
+func (m *MachineHealthCheck) ValidateCommonFields(specPath *field.Path) field.ErrorList {
```
Comment (sbueringer) [inline api/v1beta1/machinehealthcheck_webhook.go:149]: Should/can we add a disclaimer that this func is only used for our internal validation and is not supposed to be used by external consumers?

File: api/v1beta1/machinehealthcheck_webhook.go [hunk 4]
```diff
@@ -136,15 +136,27 @@ func (m *MachineHealthCheck) validate(old *MachineHealthCheck) error {
...
+	allErrs = append(allErrs, m.ValidateCommonFields(specPath)...)
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("MachineHealthCheck").GroupKind(), m.Name, allErrs)
+}
+
+// ValidateCommonFields validates UnhealthyConditions NodeStartupTimeout, MaxUnhealthy, and RemediationTemplate of the MHC.
+// It is used to help validate other types which define MachineHealthChecks such as MachineHealthCheckClass and MachineHealthCheckTopology.
```
Comment (sbueringer) [inline api/v1beta1/machinehealthcheck_webhook.go:148]:
```suggestion
// It is also used to help validate other types which define MachineHealthChecks such as MachineHealthCheckClass and MachineHealthCheckTopology.
```
nit (given it's also used for MHC itself)

File: api/v1beta1/machinehealthcheck_webhook.go [hunk 5]
```diff
@@ -165,8 +176,12 @@ func (m *MachineHealthCheck) validate(old *MachineHealthCheck) error {
 		)
 	}
 
-	if len(allErrs) == 0 {
-		return nil
+	if len(m.Spec.UnhealthyConditions) == 0 {
+		allErrs = append(allErrs, field.Forbidden(
+			specPath.Child("unhealthyConditions"),
+			"must have at least one value",
```
Comment (sbueringer) [inline api/v1beta1/machinehealthcheck_webhook.go:182]:
```suggestion
			"must have at least one entry",
```
up to you, not sure what sounds/fits better

File: api/v1beta1/machinehealthcheck_webhook.go [hunk 6]
```diff
@@ -165,8 +176,12 @@ func (m *MachineHealthCheck) validate(old *MachineHealthCheck) error {
 		)
 	}
 
-	if len(allErrs) == 0 {
-		return nil
+	if len(m.Spec.UnhealthyConditions) == 0 {
+		allErrs = append(allErrs, field.Forbidden(
```
Comment (sbueringer) [inline api/v1beta1/machinehealthcheck_webhook.go:180]: Should we get rid of the kubebuilder marker? Just would want to avoid duplication in general

File: internal/webhooks/cluster.go [hunk 7]
```diff
@@ -366,64 +366,64 @@ func (webhook *Cluster) getClusterClassForCluster(ctx context.Context, cluster *
...
-			))
-		}
-		// Ensure ControlPlane MachineHealthCheck defines UnhealthyConditions.
-		if len(cluster.Spec.Topology.ControlPlane.MachineHealthCheck.MachineHealthCheckClass.UnhealthyConditions) == 0 {
-			allErrs = append(allErrs, field.Forbidden(
-				field.NewPath("spec", "topology", "controlPlane", "machineHealthCheck", "unhealthyConditions"),
-				"must have at least one value",
-			))
+	if cluster.Spec.Topology.ControlPlane.MachineHealthCheck != nil {
+		path := field.NewPath("spec", "topology", "controlPlane", "machineHealthCheck")
```
Comment (sbueringer) [inline internal/webhooks/cluster.go:370]:
```suggestion
		fldPath := field.NewPath("spec", "topology", "controlPlane", "machineHealthCheck")
```
nit: for consistency

File: internal/webhooks/cluster.go [hunk 8]
```diff
@@ -366,64 +366,64 @@ func (webhook *Cluster) getClusterClassForCluster(ctx context.Context, cluster *
...
 		for i, md := range cluster.Spec.Topology.Workers.MachineDeployments {
-			// If MachineHealthCheck is defined ensure it defines UnhealthyConditions.
-			if md.MachineHealthCheck != nil && !md.MachineHealthCheck.MachineHealthCheckClass.IsZero() {
-				if len(md.MachineHealthCheck.MachineHealthCheckClass.UnhealthyConditions) == 0 {
-					allErrs = append(allErrs, field.Forbidden(
-						field.NewPath("spec", "topology", "workers", "machineDeployments", "machineHealthCheck").Index(i).Child("unhealthyConditions"),
-						"must have at least one value",
-					))
+			if md.MachineHealthCheck != nil {
+				path := field.NewPath("topology", "workers", "machineDeployments", "machineHealthCheck").Index(i)
```
Comment (sbueringer) [inline internal/webhooks/cluster.go:404]:
```suggestion
				fldPath := field.NewPath("spec","topology", "workers", "machineDeployments", "machineHealthCheck").Index(i)
```
fldPath for consistency + it looks like we lost spec

File: internal/webhooks/clusterclass.go [hunk 9]
```diff
@@ -538,33 +539,45 @@ func validateMachineHealthCheckClasses(clusterClass *clusterv1.ClusterClass) fie
 
 	// Validate ControlPlane MachineHealthCheck if defined.
 	if clusterClass.Spec.ControlPlane.MachineHealthCheck != nil {
+		controlPlaneMHCPath := field.NewPath("spec", "controlPlane", "machineHealthCheck")
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:542]: Maybe also just fldPath? This variable is scoped to the if anyway (same below for machineDeploymentMHCPath)

But if you prefer to keep it, just keep it :)

File: internal/webhooks/clusterclass.go [hunk 10]
```diff
@@ -538,33 +539,45 @@ func validateMachineHealthCheckClasses(clusterClass *clusterv1.ClusterClass) fie
...
 		if md.MachineHealthCheck == nil {
 			continue
 		}
-		if len(md.MachineHealthCheck.UnhealthyConditions) == 0 {
-			allErrs = append(allErrs, field.Forbidden(
-				field.NewPath("spec", "workers", "machineDeployments", "machineHealthCheck").Index(i).Child("unhealthyConditions"),
-				"must be defined and have at least one value",
-			))
-		}
+		machineDeploymentMHCPath := field.NewPath("spec", "machineDeployments", "machineHealthCheck").Index(i)
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:561]:
```suggestion
		machineDeploymentMHCPath := field.NewPath("spec", "workers", "machineDeployments", "machineHealthCheck").Index(i)
```

File: internal/webhooks/clusterclass.go [hunk 11]
```diff
@@ -538,33 +539,45 @@ func validateMachineHealthCheckClasses(clusterClass *clusterv1.ClusterClass) fie
...
+
+		allErrs = append(allErrs, validateMachineHealthCheckClass(machineDeploymentMHCPath, clusterClass.Namespace, md.MachineHealthCheck)...)
 	}
 	return allErrs
 }
+
+// validateMachineHealthCheckClass validates the MachineHealthCheckSpec fields defined in a MachineHealthCheckClass.
+func validateMachineHealthCheckClass(specPath *field.Path, namepace string, m *clusterv1.MachineHealthCheckClass) field.ErrorList {
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:569]:
```suggestion
func validateMachineHealthCheckClass(fldPath *field.Path, namepace string, m *clusterv1.MachineHealthCheckClass) field.ErrorList {
```
Let's also use fldPath, not sure what specPath is :)

File: api/v1beta1/machinehealthcheck_webhook.go (hunk 3 above)
Comment (sbueringer) [inline api/v1beta1/machinehealthcheck_webhook.go:149]:
```suggestion
func (m *MachineHealthCheck) ValidateCommonFields(fldPath *field.Path) field.ErrorList {
```

Comment (sbueringer) [reply api/v1beta1/machinehealthcheck_webhook.go:149]: I think it was just called specPath because it was literally "spec", let's use the more generic fldPath here

**Review by sbueringer — COMMENTED** [review]
Looks good to me in general

just a few nits

File: api/v1beta1/machinehealthcheck_webhook.go (hunk 3 above)
Comment (killianmuldoon) [reply api/v1beta1/machinehealthcheck_webhook.go:149]: Added - I guess the alternative is to add this into our /internal package if we want to really keep it out of the API? Given what it does right now I don't think it's a big issue, but that would be the 'right' solution IMO.

File: api/v1beta1/machinehealthcheck_webhook.go (hunk 6 above)
Comment (killianmuldoon) [reply api/v1beta1/machinehealthcheck_webhook.go:180]: I don't think getting rid of the marker is great as external consumers could be using this as included validation of those fields, right? I think we talked about getting rid of these before and decided against it.

Agreed on duplication though, and this is definitely something that could be done in a future api version, just not sure if it's possibly breaking with no benefit right now.

File: api/v1beta1/machinehealthcheck_webhook.go (hunk 3 above)
Comment (sbueringer) [reply api/v1beta1/machinehealthcheck_webhook.go]: I think it's fine. Considering we leak the behavior through ValidateCreate and ValidateUpdate anyway and we also kind of don't want to treat them as an API

File: api/v1beta1/machinehealthcheck_webhook.go (hunk 6 above)
Comment (sbueringer) [reply api/v1beta1/machinehealthcheck_webhook.go:181]: Let's keep the markers in the current PR to avoid discussions and delay.

I'll open an issue for it and try to get the framing right :)

Comment (sbueringer) [reply api/v1beta1/machinehealthcheck_webhook.go:181]: (just resolve)

**Comment by sbueringer** [comment]
Thx!

/lgtm

File: api/v1beta1/machinehealthcheck_webhook.go [hunk 12]
```diff
@@ -136,37 +136,53 @@ func (m *MachineHealthCheck) validate(old *MachineHealthCheck) error {
...
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("MachineHealthCheck").GroupKind(), m.Name, allErrs)
+}
+
+// ValidateCommonFields validates UnhealthyConditions NodeStartupTimeout, MaxUnhealthy, and RemediationTemplate of the MHC.
+// It is also used to help validate other types which define MachineHealthChecks such as MachineHealthCheckClass and MachineHealthCheckTopology.
+// Note: this function is used for internal validation and is not intended for external consumption.
+func (m *MachineHealthCheck) ValidateCommonFields(fldPath *field.Path) field.ErrorList {
```
Comment (fabriziopandini) [inline api/v1beta1/machinehealthcheck_webhook.go:150]: only one nit about naming because Common is relly generic
what about ValidateTopologyFields?

Also, the note is kind of confusing. if this is a public method part of the API it is hard to say not for external consumption... if we want this to be an internal method let's move it under internal/topology(/someting)

Comment (sbueringer) [reply api/v1beta1/machinehealthcheck_webhook.go:150]: To be honest with this implementation I would prefer Common because Topology doesn't sound right (does fields are not primarily fields of the Topology, they are shared between MHC and the MHC in ClusterClass and Cluster.spec.topology).

> Also, the note is kind of confusing. if this is a public method part of the API it is hard to say not for external consumption... if we want this to be an internal method let's move it under internal/topology(/someting)

Good point, but I wouldn't move it into a new package.

I played around a bit and we can just move the func into `internal/webhooks`

I open a PR on top of the PR to showcase how it woud look like

Comment (sbueringer) [reply api/v1beta1/machinehealthcheck_webhook.go:150]: Okay just realized that doesn't work because of a cyclic dependency.

So tl;dr we can't move this func into another package if we use `MachineHealthCheck` in the func signature.
api/v1beta1 => internal/webhook (to call the func)
internal/webhooks => api/v1beta1 (for the API type)

The only way to move this code out of the api package is as far as I can tell to move the entire webhook code like we did for Cluster & ClusterClass. But not sure if that's what we want to do.

The idea behind the disclaimer was just to give folks a warning that they shouldn't rely on this code (Just the same as they shouldn't really rely on ValidateCreate, ValidateUpdate, ValidateDelete on our API types).

But we can also drop the disclaimer and then we just deprecate the func if we ever want to get rid of it.

Comment (killianmuldoon) [reply api/v1beta1/machinehealthcheck_webhook.go:150]: IMO we should drop the disclaimer as it doesn't really work to create a public API but state it shouldn't be used. I think adding this function here is fine though - there's no need to really add it to an internal package given ValidateCreate etc. are exposed here.

Alternative is to make a webhook-only func which copies this one and link them with a comment / ref to ensure they're updated in line (this func probably won't be updated much)

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

File: api/v1beta1/machinehealthcheck_webhook.go (hunk 6 above)
Comment (sbueringer) [reply api/v1beta1/machinehealthcheck_webhook.go:180]: Created issue: https://github.com/kubernetes-sigs/cluster-api/issues/7634
